File: tests/fixtures/kotlin/companion_and_objects.kt
package app

/** Top-level singleton object. */
object Logger {
    private const val PREFIX = "[app]"

    fun log(msg: String) {
        println("$PREFIX $msg")
    }
}

/** Class with a named companion object that carries static-like state. */
class Registry {
    companion object Factory {
        private var counter: Int = 0
        fun next(): Int {
            counter += 1
            return counter
        }
    }

    fun id(): Int = Factory.next()
}

/** Class with an unnamed companion — Kotlin auto-names it `Companion`. */
class Cache {
    companion object {
        const val SIZE = 256
        fun empty(): Cache = Cache()
    }
}

/** Object inheriting a class and implementing an interface. */
interface Named {
    val name: String
}

open class BaseHandler {
    open fun handle() { }
}

object RootHandler : BaseHandler(), Named {
    override val name: String = "root"
    override fun handle() { }
}

File: tests/fixtures/scala/companion_and_objects.scala
package app

/** Top-level singleton object. */
object Logger {
  private val Prefix: String = "[app]"

  def log(msg: String): Unit = println(s"$Prefix $msg")
}

/**
 * Class + companion object pair — Scala's idiomatic equivalent of
 * Kotlin's companion object, but modelled as two SEPARATE top-level
 * declarations with the same name.
 */
class Registry {
  def id: Int = Registry.nextId
}

object Registry {
  private var counter: Int = 0
  def nextId: Int = {
    counter += 1
    counter
  }
}

/** Named singleton implementing a trait. */
trait Named {
  def name: String
}

class BaseHandler {
  def handle(): Unit = ()
}

object RootHandler extends BaseHandler with Named {
  override def name: String = "root"
  override def handle(): Unit = ()
}
